### PR TITLE
Kafka Modern C++ Example

### PR DESCRIPTION
**Pull request by runwuf**
Kafka Modern C++ Example uses modern-cpp-kafka

### CODE REVIEW

File: kafka/cc/avn_KafkaConsumer_Simple.cc [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+                std::cout << "    Key   [" << record.key().toString() << "]" << std::endl;
+                std::cout << "    Value [" << record.value().toString() << "]" << std::endl;
+            } else {
+                std::cerr << record.toString() << std::endl;
+            }
+        }
+    }
+
+    // No explicit close is needed, RAII will take care of it
+    consumer.close();
```
Comment (tkren) [inline kafka/cc/avn_KafkaConsumer_Simple.cc:58]:
```suggestion
    // Create a consumer instance
    // No explicit consumer.close() is needed, RAII will take care of it
    KafkaConsumer consumer(props);

    // Subscribe to topics
    consumer.subscribe({topic});

    while (running) {
        // Poll messages from Kafka brokers
        auto records = consumer.poll(std::chrono::milliseconds(100));

        for (const auto& record: records) {
            if (!record.error()) {
                std::cout << "Got a new message..." << std::endl;
                std::cout << "    Topic    : " << record.topic() << std::endl;
                std::cout << "    Partition: " << record.partition() << std::endl;
                std::cout << "    Offset   : " << record.offset() << std::endl;
                std::cout << "    Timestamp: " << record.timestamp().toString() << std::endl;
                std::cout << "    Headers  : " << toString(record.headers()) << std::endl;
                std::cout << "    Key   [" << record.key().toString() << "]" << std::endl;
                std::cout << "    Value [" << record.value().toString() << "]" << std::endl;
            } else {
                std::cerr << record.toString() << std::endl;
            }
        }
    }
```

File: kafka/cc/avn_KafkaConsumer_Simple.cc [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+                std::cout << "Got a new message..." << std::endl;
+                std::cout << "    Topic    : " << record.topic() << std::endl;
+                std::cout << "    Partition: " << record.partition() << std::endl;
+                std::cout << "    Offset   : " << record.offset() << std::endl;
+                std::cout << "    Timestamp: " << record.timestamp().toString() << std::endl;
+                std::cout << "    Headers  : " << toString(record.headers()) << std::endl;
+                std::cout << "    Key   [" << record.key().toString() << "]" << std::endl;
+                std::cout << "    Value [" << record.value().toString() << "]" << std::endl;
+            } else {
+                std::cerr << record.toString() << std::endl;
```
Comment (tkren) [inline kafka/cc/avn_KafkaConsumer_Simple.cc:52]: I guess we need to dump some sort of error message here? Maybe even bail out early?

File: kafka/cc/avn_KafkaProducer_Simple.cc [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+                std::cerr << "Message failed to be delivered: " << error.message() << std::endl;
+            }
+        };
+
+        if (running) {
+            // Send a message
+            producer.send(record, deliveryCb);
+            // Close the producer explicitly(or not, since RAII will take care of it)
+            producer.close();
+        }
```
Comment (tkren) [inline kafka/cc/avn_KafkaProducer_Simple.cc:56]: same as above

File: kafka/cc/config.hh [hunk 4]
```diff
@@ -0,0 +1,10 @@
+#include <kafka/KafkaProducer.h>
+
+extern const kafka::Properties props({  {"bootstrap.servers", {getenv("KAFKA_BROKER_LIST")}},
+                    {"security.protocol", {getenv("SECURITY_PROTOCOL")}},
+                    {"ssl.ca.location"  , {getenv("SSL_CA_LOCATION")}},
+                    {"ssl.certificate.location" , {getenv("SSL_CERTIFICATE_LOCATION")}},
+                    {"ssl.key.location"         , {getenv("SSL_KEY_LOCATION")}}
+});
```
Comment (tkren) [inline kafka/cc/config.hh:8]: "feels" better this way:
```suggestion
extern const kafka::Properties props({
                    {"bootstrap.servers", {getenv("KAFKA_BROKER_LIST")}},
                    {"security.protocol", {getenv("SECURITY_PROTOCOL")}},
                    {"ssl.ca.location"  , {getenv("SSL_CA_LOCATION")}},
                    {"ssl.certificate.location" , {getenv("SSL_CERTIFICATE_LOCATION")}},
                    {"ssl.key.location"         , {getenv("SSL_KEY_LOCATION")}}
});
```

**Review by tkren — CHANGES_REQUESTED** [review]
LGTM, I guess we should run this through some linter

File: kafka/cc/avn_KafkaProducer_Simple.cc (hunk 3 above)
Comment (runwuf) [reply kafka/cc/avn_KafkaProducer_Simple.cc:57]: producer.close takes care of errors
```
    const Error result = flush(timeout);
    if (result.value() == RD_KAFKA_RESP_ERR__TIMED_OUT)
    {
        KAFKA_API_DO_LOG(Log::Level::Notice, "purge messages before close, outQLen[%d]", rd_kafka_outq_len(getClientHandle()));
        purge();
    }
```

File: kafka/cc/avn_KafkaProducer_Simple.cc [hunk 5]
```diff
@@ -0,0 +1,63 @@
+#include <kafka/KafkaProducer.h>
+
+#include <cstdlib>
+#include <iostream>
+#include <string>
+#include <signal.h>
+
+#include "config.hh"
+
+std::atomic_bool running = {true};
```
[the remark this answers is not in the document]
Comment (runwuf) [reply kafka/cc/avn_KafkaProducer_Simple.cc:10]: @aiven-sal appreciate the feedback, I learned something new!
https://en.cppreference.com/w/cpp/atomic/atomic_flag

File: kafka/cc/avn_KafkaConsumer_Simple.cc [hunk 6]
```diff
@@ -7,16 +7,16 @@
...
-    if (sig != SIGINT) return;
+void signalHandler(int signum) {
+    if (signum != SIGINT) return;
 
-    if (running) {
-        running = false;
-    } else {
-        // Restore the signal handler, -- to avoid stuck with this handler
-        signal(SIGINT, SIG_IGN); // NOLINT
+    if (!stop.test_and_set()) {
```
[the remark this answers is not in the document]
Comment (runwuf) [reply kafka/cc/avn_KafkaConsumer_Simple.cc]: 💯 !! not sure what I was thinking... need more coffee 🤦

File: kafka/cc/ccexample.env [hunk 7]
```diff
@@ -0,0 +1,8 @@
+PROJECT="myproject"
+SERVICE="cc-example"
+export KAFKA_BROKER_LIST="cc-example-myproject.a.aivencloud.com:24949"
+export KAFKA_TOPIC="ccexample-test"
+export SECURITY_PROTOCOL="ssl"
+export SSL_CA_LOCATION="./ca.pem"
+export SSL_CERTIFICATE_LOCATION="./service.cert"
+export SSL_KEY_LOCATION="./service.key"
```
[the remark this answers is not in the document]
Comment (runwuf) [reply kafka/cc/ccexample.env]: just checked all the files again, should be all good now 🙏

**Pull request merged**